### PR TITLE
WIP: Relax restrictions on specsig for vararg

### PR DESCRIPTION
**Pull request by Keno**
Needs a few tests, but I want to run nanosoldier on this beforehand.

### CODE REVIEW

**Comment by Keno** [comment]
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/27022#issuecomment-387881735) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/42083e9670ac938698c4b173d2dbff1b6805ec77/e3c759d_vs_a0b1e98/report.md). cc @ararslan

**Comment by ararslan** [comment]
Still needs tests

**Comment by vtjnash** [comment]
This is still marked as WIP (including in the commit), that should have been removed before merging.

Anyways, why did this need to be merged right now? It now typically emits worse code, and then depends on our custom llvm-alloc-opt pass to do be able to do several SROA passes to recover. (caught me by surprise that `-O1` is now performing unnecessary allocations, so I wasted a time thinking it was somehow my fault that the SparseArray tests were failing on one of my branches)

File: src/codegen.cpp [hunk 1]
```diff
@@ -5901,17 +5896,23 @@ static std::unique_ptr<Module> emit_function(
...
-            jl_cgval_t tuple = emit_new_struct(ctx, vi.value.typ, nvargs, vargs);
-            // FIXME: this may assert since the type of vi might not be isbits here
-            emit_vi_assignment_unboxed(ctx, vi, NULL, tuple);
+            if (jl_is_concrete_type(vi.value.typ)) {
+                jl_cgval_t tuple = emit_new_struct(ctx, vi.value.typ, ctx.nvargs, vargs);
+                // FIXME: this may assert since the type of vi might not be isbits here
+                emit_varinfo_assign(ctx, vi, tuple);
+            } else {
+                jl_cgval_t tuple = mark_julia_type(ctx, emit_jlcall(ctx, prepare_call(jltuple_func), maybe_decay_untracked(V_null),
+                    vargs, ctx.nvargs), true, vi.value.typ);
```
Comment (vtjnash) [inline src/codegen.cpp]: later in this function, we may assert that `restTuple` is set thusly:
```
+                restTuple = emit_jlcall(ctx, prepare_call(jltuple_func), maybe_decay_untracked(V_null), vargs, ctx.nvargs);
+                jl_cgval_t tuple = mark_julia_type(ctx, restTuple, true, vi.value.typ);
```

File: src/codegen.cpp [hunk 2]
```diff
@@ -5901,17 +5896,23 @@ static std::unique_ptr<Module> emit_function(
...
                 bool isboxed;
                 Type *llvmArgType = julia_type_to_llvm(argType, &isboxed);
                 vargs[i - nreq] = get_specsig_arg(argType, llvmArgType, isboxed);
             }
-            jl_cgval_t tuple = emit_new_struct(ctx, vi.value.typ, nvargs, vargs);
-            // FIXME: this may assert since the type of vi might not be isbits here
-            emit_vi_assignment_unboxed(ctx, vi, NULL, tuple);
+            if (jl_is_concrete_type(vi.value.typ)) {
+                jl_cgval_t tuple = emit_new_struct(ctx, vi.value.typ, ctx.nvargs, vargs);
+                // FIXME: this may assert since the type of vi might not be isbits here
```
Comment (vtjnash) [inline src/codegen.cpp]: remove comment (since now we know that `vi.value.typ` is definitely inferred)

File: src/cgutils.cpp [hunk 3]
```diff
@@ -1585,9 +1585,15 @@ static jl_cgval_t emit_getfield_knownidx(jl_codectx_t &ctx, const jl_cgval_t &st
...
-    Value *valen = ctx.builder.CreateSub((Value*)ctx.argCount,
-                                     ConstantInt::get(T_int32, nreq));
+    Value *valen = NULL;
+    if (ctx.nvargs != -1) {
+        valen = ConstantInt::get(T_int32, ctx.nvargs);
+    } else {
+        assert(ctx.argCount);
+        int nreq = ctx.nReqArgs;
+        valen = ctx.builder.CreateSub((Value*)ctx.argCount,
+                                        ConstantInt::get(T_int32, nreq));
```
Comment (vtjnash) [inline src/cgutils.cpp]: indentation

File: src/cgutils.cpp [hunk 4]
```diff
@@ -1585,9 +1585,15 @@ static jl_cgval_t emit_getfield_knownidx(jl_codectx_t &ctx, const jl_cgval_t &st
 // emit length of vararg tuple
 static Value *emit_n_varargs(jl_codectx_t &ctx)
 {
-    int nreq = ctx.nReqArgs;
-    Value *valen = ctx.builder.CreateSub((Value*)ctx.argCount,
-                                     ConstantInt::get(T_int32, nreq));
+    Value *valen = NULL;
+    if (ctx.nvargs != -1) {
+        valen = ConstantInt::get(T_int32, ctx.nvargs);
+    } else {
```
Comment (vtjnash) [inline src/cgutils.cpp]: newline after `}` (per Jeff syntax rules)

**Comment by Keno** [comment]
Causes spurious allocations in broadcast code otherwise (on the caller side, since it needs to box everything for jlcall), unless everything is inlined just right. The new optimizer happened to push one of the functions over the inlining threshold, which started causing test failures because of the spurious allocation.

**Pull request merged**